packages-exp/auth-exp/src/core/user/token_manager.ts

```diff
@@ -99,18 +100,47 @@ export class StsTokenManager {
     return null;
   }
 
-  // TODO(lisajian): Fill in once refresh listener is added
   private async getTokenAndTriggerCallback(
-    _auth: AuthInternal,
-    _forceRefresh = false
+    auth: AuthInternal,
+    forceRefresh = false
   ): Promise<string | null> {
-    return null;
+    if (forceRefresh) {
+      // TODO(lisajian): Add and use client error code - ERROR_TOKEN_REFRESH_UNAVAILABLE
+      _assert(auth._customTokenProvider, auth, AuthErrorCode.INTERNAL_ERROR);
+      return this.refreshWithCustomToken(auth);
+    }
+
+    // isExpired includes a buffer window, during which the access token will still be returned even
+    // if a customTokenProvider is not set yet.
+    if (this.isExpired) {
+      _assert(
+        this.expirationTime && Date.now() < this.expirationTime!,
+        auth,
+        AuthErrorCode.TOKEN_EXPIRED
+      );
+      if (!auth._customTokenProvider) {
+        return this.accessToken;
+      }
+      return this.refreshWithCustomToken(auth);
+    }
+
+    return this.accessToken;
   }
 
   clearRefreshToken(): void {
     this.refreshToken = null;
   }
 
+  private async refreshWithCustomToken(
+    auth: AuthInternal
+  ): Promise<string | null> {
+    const token = await auth._customTokenProvider!.getCustomToken();
+    // signInWithCustomToken creates a new User and subsequently new StsTokenManager, so
```

Comment:
this feels like a memory leak in the making, can we not have it do that in pass through mode?

Reply:
I have an idea about how to do this... let's discuss in our meeting

Reply:
Chatted with Sam offline - there's going to be some churn on this PR, heads up to any folks that are taking a look right now, please hold off until the next revision (i.e. when there are 4 commits on this PR), thank you!

Reply:
Updated in most recent revision - moved refresh logic to custom_token and called API directly rather than signInWithCustomToken. Thanks for walking me through it Sam, and thank you Alex for highlighting the issue!